ctapipe_io_lst/calibration.py

```diff
@@ -0,0 +1,700 @@
+        Fill the r1 container.
+
+        Parameters
+        ----------
+        """
+        super().__init__(
+            subarray=subarray, config=config, parent=parent, **kwargs
+        )
+
+        self.mon_data = None
+        self.last_readout_time = {}
+        self.first_cap = {}
+        self.first_cap_old = {}
+        self.fbn = {}
+        self.fan = {}
+
+        for tel_id in self.subarray.tel:
+            shape = (N_GAINS, N_PIXELS, N_CAPACITORS_PIXEL)
+            self.last_readout_time[tel_id] = np.zeros(shape)
+
+            shape = (N_GAINS, N_PIXELS)
+            self.first_cap[tel_id] = np.zeros(shape, dtype=int)
+            self.first_cap_old[tel_id] = np.zeros(shape, dtype=int)
+
+        # set the right default for our default selector, change back afterwards
+        # to not impact other code.
+        old_default = ThresholdGainSelector.threshold.default_value
+        ThresholdGainSelector.threshold.default_value = 20.9
+        self.gain_selector = GainSelector.from_name(
+            self.gain_selector_type, parent=self
+        )
+
+        # to make sure it is accessed and the default is set before setting back
+        if isinstance(self.gain_selector, ThresholdGainSelector):
+            self.log.debug(
+                f'Using gain selection threshold: {self.gain_selector.threshold}'
+            )
+        ThresholdGainSelector.threshold.default_value = old_default
+
+        if self.calibration_path is not None:
+            self.mon_data = self._read_calibration_file(self.calibration_path)
+
+    def calibrate(self, event: ArrayEventContainer):
+        for tel_id, r0 in event.r0.tel.items():
+            r1 = event.r1.tel[tel_id]
+
+            # update first caps
+            self.first_cap_old[tel_id][:] = self.first_cap[tel_id][:]
+            self.first_cap[tel_id] = get_first_capacitor_for_modules(
+                event.lst.tel[tel_id].evt.first_capacitor_id,
+                event.lst.tel[tel_id].svc.pixel_ids,
+            )
+
+            # apply drs4 corrections
+            self.subtract_pedestal(event, tel_id)
+            self.time_lapse_corr(event, tel_id)
+            self.interpolate_spikes(event, tel_id)
+
+            # remove samples at beginning / end of waveform
+            start = self.r1_sample_start.tel[tel_id]
+            end = self.r1_sample_end.tel[tel_id]
+            r1.waveform = r1.waveform[..., start:end]
+            waveform = r1.waveform
+
+            waveform -= self.offset.tel[tel_id]
+
+            # apply monitoring data corrections
+            if self.mon_data is not None:
+                calibration = self.mon_data.tel[tel_id].calibration
+                waveform -= calibration.pedestal_per_sample[:, :, np.newaxis]
+                waveform *= calibration.dc_to_pe[:, :, np.newaxis]
+
+            waveform = waveform.astype(np.float32)
+
+            # gain selection
+            r1.selected_gain_channel = self.gain_selector(waveform)
```

Comment:
We should be able to de-select some of the steps performed here, in the case we want
1) not to calibrate at all (e.g. for the calculation of the drs4 baseline coefficients)
2) not to gain select and charge calibrate (e.g. for the estimation of the charge calibration coefficients)
3) not to gain select, time calibrate and charge calibrate (e.g. for the estimation of the drs4 time coefficients)

Reply:
Most of what you ask is already possible by just not giving the corresponding calibration file.
The only problem is making gain selection optional.

It is easy to implement that here, but ctapipe currently does not handle R1 with two gains, so I'd guess this option would be usable only for lstchain (which is fine, it is where the calib files are calculated).

However, I think since the R1 data model no contains "long events" with both gain channels, ctapipe also has to support this.

Anyways, for the time, I'll add an option to disable gain selection.

Reply:
Actually, "long events" are something completely different, sorry for mixing this up.

Reply:
> Most of what you ask is already possible by just not giving the corresponding calibration file.

Perfect.

> The only problem is making gain selection optional.
> 
> It is easy to implement that here, but ctapipe currently does not handle R1 with two gains, so I'd guess this option would be usable only for lstchain (which is fine, it is where the calib files are calculated).

That R1 has only one channel is new for me (it was decided recently?), the calibration code has been developed using the two gains in the R1 level

> Anyways, for the time, I'll add an option to disable gain selection.

Yes, this will be sufficient for now.